packages/multichain/src/middlewares/multichainMethodCallValidator.ts

```diff
@@ -0,0 +1,114 @@
+import { makeCustomResolver } from '@open-rpc/schema-utils-js/build/parse-open-rpc-document';
+import type { Schema, ValidationError } from 'jsonschema';
+import { Validator } from 'jsonschema';
+
+const transformError = (
+  error: ValidationError,
+  param: ContentDescriptorObject,
+  got: unknown,
+) => {
+  // if there is a path, add it to the message
+  const message = `${param.name}${
+    error.path.length > 0 ? `.${error.path.join('.')}` : ''
+  } ${error.message}`;
+
+  return {
+    code: -32602, // TODO: could be a different error code or not wrapped in json-rpc error, since this will also be wrapped in a -32602 invalid params error
+    message,
+    data: {
+      param: param.name,
+      path: error.path,
+      schema: error.schema,
+      got,
+    },
+  };
+};
+
+const v = new Validator();
+
+const dereffedPromise = dereferenceDocument(
+  MultiChainOpenRPCDocument as unknown as OpenrpcDocument,
+  makeCustomResolver({}),
+);
+const multichainMethodCallValidator = async (
+  method: string,
+  params: JsonRpcParams | undefined,
+) => {
+  const dereffed = await dereffedPromise;
+
+  const methodToCheck = dereffed.methods.find(
+    (m) => (m as unknown as ContentDescriptorObject).name === method,
+  );
+
+  const errors: JsonRpcError[] = [];
+  if (
+    !methodToCheck ||
+    !isObject(methodToCheck) ||
+    !('params' in methodToCheck)
+  ) {
+    return [
+      {
+        code: -32601,
+        message: 'The method does not exist / is not available.',
+        data: {
+          method,
+        },
+      },
+    ];
+  }
+  // check each param and aggregate errors
+  (methodToCheck as unknown as MethodObject).params.forEach((param) => {
+    const p = param as ContentDescriptorObject;
+    if (!isObject(params)) {
+      errors.push({
+        code: -32602,
+        message: 'Invalid method parameter(s).',
+      });
+      return;
+    }
+    const paramToCheck = params[p.name];
+
+    const result = v.validate(paramToCheck, p.schema as unknown as Schema, {
+      required: p.required,
+    });
+    if (result.errors) {
+      errors.push(
+        ...result.errors.map((e) => {
+          return transformError(e, p, paramToCheck) as JsonRpcError;
+        }),
+      );
+    }
+  });
+  if (errors.length > 0) {
+    return errors;
+  }
+  // feels like this should return true to indicate that its valid but i'd rather check the falsy value since errors
+  // would be an array and return true if it's empty
+  return false;
+};
+
+export const multichainMethodCallValidatorMiddleware: JsonRpcMiddleware<
+  JsonRpcParams,
+  Json
+> = function (request, _response, next, end) {
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  multichainMethodCallValidator(request.method, request.params).then(
+    (errors) => {
+      if (errors) {
+        return end(rpcErrors.invalidParams<JsonRpcError[]>({ data: errors }));
+      }
+      return next();
+    },
+  );
+};
```

Comment:
Why can't this be an async middleware? Then we can await this and avoid the floating promise:

``` suggestion
export const multichainMethodCallValidatorMiddleware = createAsyncMiddleware(
  async (request, _response, next) => {
    const errors = await multichainMethodCallValidator(
      request.method,
      request.params,
    );
    if (errors) {
      throw rpcErrors.invalidParams<JsonRpcError[]>({ data: errors });
    }
    return await next();
  },
);
```